src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_updates_management.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * @file adu_updates_management.c
+ * @brief Implementation for the deployments and updates management related functions.
+ *
+ * @copyright Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ */
+
+#include "aduc/adu_updates_management.h"
+#include "aduc/adu_mosquitto_utils.h"
+#include "aduc/adu_mqtt_protocol.h"
+#include "aduc/logging.h"
+
+#include <aducpal/time.h> // time_t
+#include <mosquitto.h> // mosquitto related functions
+#include <mqtt_protocol.h> // mosquitto_property
+
+typedef enum ADU_MQTT_UPDATE_SYNC_STATE_TAG
+{
+    ADU_MQTT_UPDATE_SYNC_STATE_FAILED = -2,
+    ADU_MQTT_UPDATE_SYNC_STATE_PROCESS_FAILED = -2,
+    ADU_MQTT_UPDATE_SYNC_STATE_REQUEST_FAILED = -1,
+    ADU_MQTT_UPDATE_SYNC_STATE_UNKNOWN = 0,
+    ADU_MQTT_UPDATE_SYNC_STATE_SENDING = 1,
+    ADU_MQTT_UPDATE_SYNC_STATE_PUBLISHED = 2,
+    ADU_MQTT_UPDATE_SYNC_STATE_RESPONDED = 3,
+    ADU_MQTT_UPDATE_SYNC_STATE_PROCESSING = 4,
+    ADU_MQTT_UPDATE_SYNC_STATE_PROCESS_COMPLETED = 5,
+} ADU_MQTT_UPDATE_SYNC_STATE;
+
+typedef struct ADU_UPDATES_MANAGEMENT_STATE_TAG
+{
+    ADU_MQTT_UPDATE_SYNC_STATE upd_req_state;
+    char* upd_req_correlation_id;
+    time_t upd_req_attempt_time;
+    time_t next_upd_req_attempt_time;
+    time_t last_upd_req_success_time;
+    time_t last_upd_req_error_time;
+
+    time_t upd_resp_time;
+    char* upd_resp_content;
+
+    time_t last_upd_cn_time;
+    char*
+        last_upd_cn_correlation_id; //!< The correlation ID for the last 'update change notification'. If this value is not NULL, the client will resent an 'upd_req' message with this correlation ID.
+} ADU_UPDATES_MANAGEMENT_STATE;
```

Comment:
struct declarations should be moved out to .h to enable unit testing and should avoid static variables as there could be multiple instances, one for each module that links to this target.